### DISCUSSION TITLE
Running Ripgrep on files

### DISCUSSION THREAD

**Discussion by limaceous-bushwhacker** [junegunn/fzf#3710] [Q&A] [answered]
How can I configure a keybinding in fzf to run ripgrep over a selected array of files? I want to use fzf as a file picker to open the files in vim by default, but on pressing Ctrl+R, I want to run ripgrep over those files to search for a keyword.

Related(?): https://github.com/junegunn/fzf/blob/master/ADVANCED.md#using-fzf-as-interactive-ripgrep-launcher

**Comment by LangLangBart** [comment, marked as answer]
## Two Ideas

### A: Direct Approach

We attempt to map the capability of searching a specified number of files with ripgrep to a single
hotkey. This approach sacrifices readability, and one must navigate the complexities of escaping and
quoting.

```bash
RG_PREFIX=$'rg --with-filename --column --line-number --no-heading --color=always --smart-case ${FZF_QUERY:-\'""\'} -- $(while IFS= read -r line; do printf "%q " "$line"; done < /tmp/rg_files) || true'
fzf \
  --ansi \
  --multi \
  --delimiter : \
  --bind 'start:execute-silent(: >/tmp/rg_files)+unbind:change' \
  --bind "ctrl-r:rebind(change)+unbind(ctrl-r)+disable-search+clear-query+execute-silent(cp {+f} /tmp/rg_files)+change-preview(bat --color=always {1} --highlight-line {2})+change-preview-window(up,60%,border-bottom,+{2}+3/3,~3)+reload:eval $RG_PREFIX" \
  --bind "change:reload:sleep 0.1; eval $RG_PREFIX" \
  --bind "enter:transform:[[ {} == *:*:*:* ]] &&
    echo 'become:vim \{1} +\{2}' ||
    echo 'become:vim \{}'"
```




<img src="https://github.com/junegunn/fzf/assets/92653266/5aa3c285-afd2-4fbf-b2ea-7250dd31ec31" width="800">

---

### B: Script Approach


This idea calls a script from our normal usage. Two things are needed:

1. A modified version of the function from your link[^1] that can take a file with listed filenames
   as input. Ensure it is executable `chmod +x`, and include the directory where this file is
   located in your `$PATH`. We'll name this file `modified_rfv`.

```bash
#!/usr/bin/env bash

# Original: https://github.com/junegunn/fzf/blob/master/ADVANCED.md#using-fzf-as-interactive-ripgrep-launcher
# Modified version

while getopts :f: cmd_arg; do
  case "$cmd_arg" in
    f) file_list=$(while IFS=$'\n' read -r line; do printf "%q " "$line"; done <"${OPTARG}") ;;
    *)
      exit 1
      ;;
  esac
done
shift "$((OPTIND - 1))"

RG_PREFIX="rg --with-filename --column --line-number --no-heading --color=always --smart-case "
INITIAL_QUERY="${*:-}"
: | fzf \
  --ansi --disabled --query "$INITIAL_QUERY" \
  --bind "start:reload:$RG_PREFIX {q} -- $file_list " \
  --bind "change:reload:sleep 0.1; $RG_PREFIX {q} -- $file_list || true" \
  --delimiter : \
  --preview 'bat --color=always {1} --highlight-line {2}' \
  --preview-window 'up,60%,border-bottom,+{2}+3/3,~3' \
  --bind 'enter:become:vim {1} +{2}'
```

2. We assign the <kbd>⌃ Control</kbd> + <kbd>R</kbd> hotkey to invoke the `modified_rfv` script,
   taking a file as input. The `{+f}` represents all selected entries, which are stored in a temporary file.
   For more details, refer to `man fzf | less --pattern 'with f flag'`. One might want to use
  `execute:modified_rfv -f {+f}` to be able to return to the original list with <kbd>⎋ Escape</kbd>.

```bash
fzf \
  --multi \
  --bind 'ctrl-r:become:modified_rfv -f {+f}' \
  --bind 'enter:become:vim {+}'
```

<img src="https://github.com/junegunn/fzf/assets/92653266/c9666e44-3bbd-408b-9986-0f1bb3a75905" width="800">

---

> [!NOTE]
> The main challenge with any version was to ensure it also worked for filenames containing spaces.

[^1]: https://github.com/junegunn/fzf/blob/master/ADVANCED.md#using-fzf-as-interactive-ripgrep-launcher

**Reply by limaceous-bushwhacker** [reply]
Hey, thanks for getting back to me. It seems I might not have been clear in my question. What I'm looking for is to start with fzf as a file picker, so when I hit Enter, it opens the selected file in $EDITOR. Then, if I press Ctrl+R, I want to switch to ripgrep mode to search through those same files. Your solution seems to focus on running ripgrep over selected files, which is not exactly what I had in mind.
